Use GResource instead of deprecated GdkPixdata, when possible

nuko8 opened the pull request:
Following the instruction of

https://developer.gnome.org/gdk-pixbuf/2.31/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-new-from-inline,

replace the deprecated function `gdk_pixbuf_new_from_inline()` with
`gdk_pixbuf_new_from_resource()` to make use of GResource.

To do this,
- Restore icon image files from the icon data inlined in
  `pixmaps/stock_icons.h`, and place them in the directory `pixmaps`.
- Add an XML file which enumerates those icon files.
- Modify the build system to use `glib-compile-resources`(1) to create a C
  file from the XML file.  The C file is used to incorporate the
  information of the XML file into the Vim source.
- Modify the implementation of `gui_gtk_register_stock_icons()` in
  `gui_gtk.c` so that it will use the GResource facility.
- The last two modifications are made enable only if gdk-pixbuf >= 2.32;
  otherwise, the current implementation is used.

---

Hopefully, this PR gives one of the ways to fix the issue raised by Tony Mechelynck, e.g., https://groups.google.com/forum/#!topic/vim_dev/MLL4NhDWuGc

Best regards,
Kazunobu Kuriyama

brammool commented:
Kazunobu Kuriyama wrote:

> Following the instruction of
> 
> https://developer.gnome.org/gdk-pixbuf/2.31/gdk-pixbuf-Image-Data-in-Memory.html#gdk-pixbuf-new-from-inline,
> 
> replace the deprecated function `gdk_pixbuf_new_from_inline()` with
> `gdk_pixbuf_new_from_resource()` to make use of GResource.
> 
> To do this,
> - Restore icon image files from the icon data inlined in
>   `pixmaps/stock_icons.h`, and place them in the directory `pixmaps`.
> - Add an XML file which enumerates those icon files.
> - Modify the build system to use `glib-compile-resources`(1) to create a C
>   file from the XML file.  The C file is used to incorporate the
>   information of the XML file into the Vim source.
> - Modify the implementation of `gui_gtk_register_stock_icons()` in
>   `gui_gtk.c` so that it will use the GResource facility.
> - The last two modifications are made enable only if gdk-pixbuf >= 2.32;
>   otherwise, the current implementation is used.
> 
> ---
> 
> Hopefully, this PR gives one of the ways to fix the issue raised by Tony Mechelynck, e.g., https://groups.google.com/forum/#!topic/vim_dev/MLL4NhDWuGc

Thanks, that is quite a big change.  Will this still work with older GTK
versions?

## 

Microsoft says that MS-Windows is much better for you than Linux.
That's like the Pope saying that catholicism is much better for
you than protestantism.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\
\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\            help me help AIDS victims -- http://ICCF-Holland.org    ///

nuko8 commented:
> Will this still work with older GTK versions?

Yes, it will.

With the current implementation, the GUI obtains icon data through `GtkIconFactory`, and the latter gets them through `GtkIconSet`.

`GtkIconSet` only knows icon data in the form of `GdkPixbuf`, and therefore needn’t know where the original data come from and how they are converted into `GdkPixbuf` data.

IOW, for the GTK layer, it doesn’t matter which function, `*_from_inline()` or `*_from_resource()`, is used.

This PR honors that wonderful indirection mechanism, which makes it GTK-version independent.

## 

UPDATE:

> that is quite a big change. 

Most of it is for `GResource` and `GdkPixbuf`.   As I wrote above, the GTK layer will never know what is changed with this PR :-)

nuko8 commented:
Updated.

In addition to minor fixes and changes, there’s a substantial change in this update.

By default, `glib-compile-resources` yields C source code making use of `__attribute__((constructor))` and `__attribute__((destructor))` (and the equivalent for Windows) to automatically register and unregister GResource, respectively.

Obviously, the code won’t be compiled in case the compiler to be used doesn’t support those attributes, although the GResource manual says the default code is OK for Linux, OS X and Windows.

With this update, the option `--manual-register` is passed to `glib-compile-resources` and some code is added to `gui_gtk_x11.c` to register and unregister GResource manually, so that the code automatically generated by `glib-compile-resources` can be compiled on platforms other than Linux, OS X and Windows.

nuko8 committed:
Restore icon image files from `pixmaps/stock_icons.h`

In order to replace deprecated `gdk_pixmap_new_from_inline()` with
`gdk_pixmap_new_from_resource()`, we must first restore icon image files
from `pixmaps/stock_icons.h` in which the icon images are inlined with
deprecated `GdkPixdata`.

nuko8 committed:
Use GResource instead of GdkPixdata, if possible

- Modify the build system so that icon data to be used will be switched
  from GdkPixdata to GResource if the version of `gdk-pixbuf` is 2.32 or
  later, and if `glib-compile-resources` is in PATH.
- Add a new XML file `src/gui_gtk.gresources.xml` so that the GResource
  system can refer to the icon image files restored from
  `pixmaps/stock_icons.h`.
- Add new targets to `src/Makefile` to yield a C source file containing
  inlined GResource data, using `glib-compile-resources`.
- Add a compilation path to `src/gui_gtk_x11.c` for
  registering/unregistering the GResource system.
- Add a compilation path to `src/gui_gtk.c` for switching from
  deprecated `gdk_pixbuf_new_from_inline()` to
  `gdk_pixbuf_new_from_resource()`.
